autode/opt/optimisers/base.py

```diff
@@ -30,14 +34,24 @@ def converged(self) -> bool:
     def last_energy_change(self) -> PotentialEnergy:
         """The energy change on between the final two optimisation cycles"""
 
+    def run(self, *args: Any, **kwargs: Any) -> None:
+        raise NotImplementedError
+
+    def save(self, filename: str) -> None:
+        raise NotImplementedError
+
+    @property
+    def final_coordinates(self):
+        raise RuntimeError("A NullOptimiser has no coordinates")
```

Comment:
This is the base class, does it need RuntimeError, or should it be abstractmethod instead?

Reply:
yeah, it really should be an abstract method but there are implementation where it'd just be implemented as raising an exception so decided it was more DRY to have it in the base class. probably an argument for using more composition but hey ho!